Feature(bot): inline-button for registrations

Hrushon opened the pull request:
Added `InlineButton` and `MenuButton` to call registration process template.

https://www.notion.so/Bot-cae1343ec01d42a897219b42f3add7ae

AntonRev reviewed (commented):
* Некоторые  сворачивают кнопку и не могут потом ее найти, поэтому нужны еще инлайн кнопка и в меню.  
* Инлайн кнопка у тебя выдается обрезанная. 
* В меню можно добавлять без FotFather https://docs.python-telegram-bot.org/en/stable/telegram.menubuttoncommands.html
![2023-03-28_15-29-16](https://user-images.githubusercontent.com/113011812/228239634-595283e4-9a93-42f2-8fff-9ada866e3800.png)

AntonRev reviewed (commented):
Мне кажется добавление кнопок везде звучало лучше чем получилось. Может стоит убрать в меню? @Ivan-Skvortsov  @ADesBiysk

ADesBiysk commented:
> Мне кажется добавление кнопок везде звучало лучше чем получилось. Может стоит убрать в меню? @Ivan-Skvortsov @ADesBiysk

Мне inline кнопка не очень нравится. Я снял видео и задал вопрос в coreteam.

Ivan-Skvortsov commented:
Насколько я понял из доков телеги, для вебапп, запущенного из инлайн кнопки и из меню можно использовать [answerWebAppQuery ](https://core.telegram.org/bots/api#answerwebappquery).
Разница в том, что sendData - это js-функция из пакета телеги [telegram-web-app.js](https://telegram.org/js/telegram-web-app.js) и её можно дернуть на фронте, а answerWebAppQuery - это метод бота, который нужно дергать на бэке.
Получается, нам нужен дополнительный эндпойнт на бэке, на который будет получать данные из вебапп.

Если резюмировать, то порядок действий для получения данных из вебапп должен быть такой:
1) Получаем данные из [initData ](https://core.telegram.org/bots/webapps#initializing-web-apps:~:text=Description-,initData,-String)при помощи кастомного js-скрипта.
2) При помощи того же скрипта отправляем данные на бэк на наш эндпойнт
3) На бэке обрабатываем данные, как нам необходимо и вызываем answerWebAppData, чтобы отправить сообщение пользователю (тут нужно использовать web_app_query_id, чтобы не промахнуться с чатом)

P.S.: по сути, даже в answerWebAppQuery нет нужды, мы просто можем повесить на мейн баттон вебаппа скрипт, который выполняет п.п.1 и 2 и закрывает вебапп. А сообщение пользователю отправляем уже с бэка, в зависимости от результатов обработки

P.P.S.: ну и чтобы не делать дополнительные шаблоны и логику на бэке, есть смысл переделать и также и текущую реализацию.

Могу где-то ошибаться, поправьте меня)

Hrushon commented:
Ребята, подскажите @Ivan-Skvortsov @ADesBiysk  @AntonRev 
То, что сейчас реализовано, я так понимаю - никого не устраивает?) Нужно переделывать?

Я просто не знаю js и поэтому очень много времени трачу на него :( не хотелось бы зависать над решением, которое не пригодится впоследствии)

ADesBiysk commented:
> Ребята, подскажите @Ivan-Skvortsov @ADesBiysk @AntonRev То, что сейчас реализовано, я так понимаю - никого не устраивает?) Нужно переделывать?
> 
> Я просто не знаю js и поэтому очень много времени трачу на него :( не хотелось бы зависать над решением, которое не пригодится впоследствии)

Сейчас реализовано не так как хотелось бы. Пример, как должно работать можно посмотреть в этом боте: https://t.me/DurgerKingBot
Если данная задача требует много времени и усилий, то лучше взять другую и не мучить себя.

Hrushon commented:
Понял, буду переделывать

Hrushon committed:
feat(bot): inline- & menu-button for registrations

Version #2.
Added InlineButton & MenuButton to call registration process template.
Added new handlers.

Hrushon committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into feature/new_registration_buttons

Hrushon committed:
feat(bot): inline- & menu-button for registrations

Version # 2.
Added InlineButton & MenuButton to call registration process template.
Added new handlers.
Temporarily commented out checking of the secret token in the `webhook` endpoint.

Hrushon committed:
feat(bot): inline- & menu-button for registrations

Version # 2.
Added InlineButton & MenuButton to call registration process template.
Added new handler.
Temporarily commented out checking of the secret token in the `webhook` endpoint.

Hrushon commented:
@Ivan-Skvortsov @ADesBiysk @AntonRev

Понятно, что это не окончательный вариант, посмотрите, что выходит. Так пойдет?

1. Закомментировал проверку секретного ключа в вебхуке телеграм бота на время, чтоб проходили POST-запросы с шаблона регистрации. Пока не знаю - как в шаблоне секретный ключ получить.
2. Вроде все неплохо, единственное, не очень смотрится, что сообщение метод `answerWebAppQuery` от имени пользователя отправляет.
3. Может объединим обратно методы `update_user_data` и `register_user`? Столько повторяется там кода...

Ivan-Skvortsov commented on src/api/routers/telegram_webhook.py:
```diff
@@ -51,12 +51,13 @@ def get_register_form() -> Iterator[bytes]:
         response_description="Обновления получены",
     )
-    async def get_telegram_bot_updates(request: Request) -> dict:
+    async def get_telegram_bot_updates(request: Request) -> None:
```
Этот эндпойнт не нужно модифицировать, он должен использоваться только серверами телеги для отправки обновлений для нашего бота) 
Просто создай свой (новый) эндпойнт, предназначенный для обработки данных из вебапп. Ты же обращаешься к нему из js-скрипта и можешь сделать его таким, как тебе удобно)

Hrushon replied:
хорошо, понял

Hrushon replied:
@Ivan-Skvortsov единственное, мне кажется в этой функции ошибка - если она добавила данные в очередь бота `update_queue` зачем она возвращает словарь?

Ivan-Skvortsov commented on src/templates/registration/registration.html:
```diff
@@ -452,8 +452,21 @@
+          }
+        });
+      } else {
+        tg.sendData(JSON.stringify(data));
```
Предлагаю от sendData вообще отказаться и все кейсы (кейбоард кнопка, инлайн кнопка и меню) обрабатывать одинаково.

Ivan-Skvortsov replied:
Если это возможно, конечно)

Hrushon replied:
я не могу отказаться от функции `web_app_data` по соображениям, которые описаны в ответе к последнему замечанию поэтому ничего не изменится. только эта js функция станет больше.

Hrushon replied:
скорее невозможно, так как данные о чате и пользователе отсутствуют  при активации формы регистрации с keyboard button в `initDataUnsafe`, а больше и взять то негде

Ivan-Skvortsov replied:
Неприятно, конечно)
Получается, нам нужно либо отказываться от кейбоард кнопки, либо писать отдельно логику для кейбоард и отдельно для инлайн/меню?

Hrushon replied:
почему отказываться? 
сейчас же работает. чуток совсем добавилось логики для инлайн/меню кнопок.
для пользователя разница лишь в способе оповещения о получении данных с формы: при активации с кейборд кнопки - встроенное телеграмовское оповещение, а при активации с меню/инлайн кнопки - отправка ботом сообщения от имени пользователя в чат.

Ivan-Skvortsov commented on src/bot/handlers.py:
```diff
@@ -145,6 +195,47 @@ async def web_app_data(update: Update, context: CallbackContext) -> None:
+    )
+
+
+async def get_answer_web_app(update: Update, context: CallbackContext) -> None:
```
Эта функция нигде не используется.

Hrushon replied:
да, невнимательность, забыл удалить

Ivan-Skvortsov commented on src/bot/main.py:
```diff
@@ -41,6 +45,13 @@ def create_bot() -> Application:
         .persistence(persistence=bot_persistence)
         .build()
     )
+    bot_instance.add_handler(
+        TypeHandler(
+            dict,
+            get_web_app_query_data,
+        )
+    )
+    bot_instance.add_handler(MessageHandler(Regex(RE_PATTERN), web_app_data))
```
Довольно запутанно получилось, на мой взгляд)

Если я правильно понял назначение всех этих хэндлеров - ты пытаешься работать вокруг бота/чата телеги.
Попробуй посмотреть на эту задачу с другой стороны. 
Фактически, у тебя есть полноценный фронт (вебапп) и полноценный бэкенд (собственно, наше приложение). При этом в бэкенде у тебя есть доступ к боту (bot_instance, который лежит в app.state). 
То есть, для обработки данных из вебапп, чат телеги тебе не нужен вообще. Просто работай с ними как с любыми другими данными)

Вот, например, я накидал тестовый эндпойнт, который принимает запрос с данными из вебапп и отправляет пользователю инлайн сообщение, что все ок (взял кусок из твоего кода).
<img width="631" alt="image" src="https://user-images.githubusercontent.com/83599925/233653440-719df33c-e595-41fd-a69f-b52bbabcb412.png">

Hrushon replied:
@Ivan-Skvortsov 
если это для того, чтобы только избавиться от функции `get_web_app_query_data`, то да, спасибо, попробую сделать так.

Hrushon replied:
@Ivan-Skvortsov эта функция будет срабатывать получается вразрез с очередью бота `update_queue`, не могу оценить - насколько это критично? ощущение, что это не очень хорошо

Ivan-Skvortsov replied:
Мне кажется, в данном случае нам это неважно. Я думаю, работу с вебапп нужно рассматривать как работу бэка с фронтом)
Могу ошибаться, конечно)

Hrushon committed:
feat(bot): inline- & menu-button for registrations

Version # 3.
Added InlineButton & MenuButton to call registration process template.
Added new handler.
Added new endpoint for a requests with data coming from registration form launched via the keyboard button.

Hrushon committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into feature/new_registration_buttons

Hrushon committed:
feat(bot): inline- & menu-button for registrations

Version # 4.
Added InlineButton & MenuButton to call registration process template.
Added new handler.
Added new endpoint for a requests with data coming from registration form launched via the keyboard button.
Changed the function of the registration template rendering in terms of adding a url from the settings-file to receive requests with data launched via `inline/menu` button.

Hrushon committed:
Merge branch 'develop' into feature/new_registration_buttons

Hrushon committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into feature/new_registration_buttons

Ivan-Skvortsov commented on src/bot/handlers.py:
```diff
@@ -76,6 +81,23 @@ async def register_user(
             )
         ),
     )
+    await context.bot.set_chat_menu_button(
```
Повторяющиеся куски кода (с кнопками) хорошо бы вынести в отдельную функцию (или функции)

Hrushon committed:
feat(bot): inline- & menu-button for registrations

Version # 5.
Added InlineButton & MenuButton to call registration process template.
Added new handler.
Added new endpoint for a requests with data coming from registration form launched via the keyboard button.
Changed the function of the registration template rendering in terms of adding a url from the settings-file to receive requests with data launched via `inline/menu` button.

Hrushon committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into feature/new_registration_buttons

Hrushon committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into feature/new_registration_buttons

Hrushon committed:
feat(bot): inline- & menu-button for registrations

Version # 6.
Added InlineButton & MenuButton to call registration process template.
Added new handler.
Added new endpoint for a requests with data coming from registration
form launched via the keyboard button.
Changed the function of the registration template rendering in terms of
adding a url from the settings-file to receive requests with data
launched via `inline/menu` button.
